### ISSUE TITLE
Magnet link doesn't start torrent download

### ISSUE THREAD

**Issue by srirambv** [brave/browser-laptop#5981]
**Did you search for similar issues before submitting this one?**
Yes

**Describe the issue you encountered:**
Magnet link doesn't start torrent download

**Expected behavior:**
Should load the torrent viewer and start downloading 

- Platform (Win7, 8, 10? macOS? Linux distro?):
Windows 10 x64

- Brave Version:
0.13.0 preview 2

- Steps to reproduce:
    1. Click on any of the magnet links from https://codepen.io/ferossity/full/qaezaB/
    2. Torrent viewer doesn't load and download doesn't start

- Screenshot if needed:

- Any related issues:
#5945

**Comment by bbondy** [comment]
@dcposch if you still have cycles available it'd be helpful if you could take a look on the chromium54 branch.

**Comment by dcposch** [comment]
@bbondy sure, checking it out

**Comment by dcposch** [comment]
Just reproduced it:

```
Uncaught SyntaxError: Identifier 'webtorrentEntryPage' has already been declared at webtorrent.js:11
```

Here's `webtorrent.js` on the Chromium54 branch:

```js
  let webtorrentEntryPage = 'gen/webtorrentPage.entry.js'

  var queryString = window.location.search
  var portMatch = queryString && queryString.match(/devServerPort=([0-9]+)/)
  var devServerPort
  if (portMatch) {
    devServerPort = portMatch[1]
  }

  let webtorrentEntryPage = 'gen/webtorrentPage.entry.js'
  // [...]
```

Looks like this commit duplicates `webtorrentEntryPage`: https://github.com/brave/browser-laptop/commit/eb19c308523b145a2962fc7d64fb44f1e1504d7f

It's on `chromium54` and a few other branches, but not on master.

**Comment by dcposch** [comment]
@bridiver

**Comment by bridiver** [comment]
looks like a rebase issue. The webtorrent.js file shouldn't exist anymore

**Comment by bridiver** [comment]
had to sort some things out in master and I'll update the changes for 54

**Comment by dcposch** [comment]
@bbondy @bridiver LMK if there's anything you'd like me to do.

Right now, Torrent Viewer doesn't work on the `chromium54` branch. It should work again if you revert https://github.com/brave/browser-laptop/commit/eb19c308523b145a2962fc7d64fb44f1e1504d7f or remove the extra `let` statement. Works on `master`.

**Comment by dcposch** [comment]
aha, looks like it's called `chrome.ipcRenderer` now

**Comment by ayumi** [comment]
@dcposch I get a CSP error when opening this magnet link in Brave 0.13.0 pre4:

![screen shot 2016-12-16 at 11 37 37](https://cloud.githubusercontent.com/assets/521861/21275950/128c2b06-c384-11e6-828b-e3c9a7bc82e8.png)

**Comment by srirambv** [comment]
Doesn't load the Torrent viewer interface on Windows x64 (0.13.0 Preview 4)
![webtorrent](https://cloud.githubusercontent.com/assets/17010094/21280636/12f363a6-c3df-11e6-929b-df96e00c0de4.gif)

**Comment by bridiver** [comment]
the urls work with open in new tab now, but do nothing if you just click on them

**Comment by srirambv** [comment]
Doesn't work even with "open in new tab". Here's the screen grab
![webtorrent1](https://cloud.githubusercontent.com/assets/17010094/21429733/076b80b8-c858-11e6-88c8-fa51223b2459.gif)

![image](https://cloud.githubusercontent.com/assets/17010094/21429705/f2ff7918-c857-11e6-9d26-12a88f8447ed.png)

**Comment by jonathansampson** [comment]
I see the same thing as @srirambv.

**Comment by bridiver** [comment]
reverting https://github.com/brave/browser-laptop/commit/eb19c308523b145a2962fc7d64fb44f1e1504d7f is definitely not the right thing to do. The original code incorrectly bypassed the webpack dev server and the csp issues should be resolved in extensions.js

**Comment by dcposch** [comment]
@bridiver sure, but https://github.com/brave/browser-laptop/commit/eb19c308523b145a2962fc7d64fb44f1e1504d7f doesn't actually run at all due to the syntax error (duplicate `let`).

not sure how that got merged.

running without the webpack dev server means you have to run `npm run webpack` after changes to the torrent extension--i don't think it affects the production build. i can try to get the Torrent Viewer working with the dev server later today to make future development faster.

what are the remaining CSP issues?

**Comment by bridiver** [comment]
actually https://github.com/brave/browser-laptop/commit/eb19c308523b145a2962fc7d64fb44f1e1504d7f was already replaced by other code to handle the dev server and that file doesn't even exist anymore

**Comment by bridiver** [comment]
there are a lot of comments on this thread that are no longer relevant so I'm not sure what the outstanding issues are, but the code from https://github.com/brave/browser-laptop/commit/eb19c308523b145a2962fc7d64fb44f1e1504d7f is definitely not part of the problem because that file was removed a while ago and the webpack dev server is handled in filtering.js

**Comment by dcposch** [comment]
i'm building brave now to try out:
- clicking a magnet link
- pasting a magnet link into the URL bar
- checking for CSP errrors

is `master` the right branch to be on?

**Comment by bridiver** [comment]
yep, everything from chromium54 has been merged and `app/extensions/torrent/js/webtorrent.js` doesn't exist there. I'm pretty sure that change was already superseded by the filtering change before it was even merged into master.

**Comment by bbondy** [comment]
I don't think the webtorrent filtering code is seeing the magnet protocol URLs at all

**Comment by bridiver** [comment]
@bbondy that is correct in some cases, but not others and I'm not sure why. Cmd-click seems to work, but normal links don't. Maybe they aren't registering as a navigation?

**Comment by srirambv** [comment]
Seeing this error on 0.13.0 Preview 7 

![image](https://cloud.githubusercontent.com/assets/17010094/21652088/26e35466-d2a3-11e6-926a-223aa2cd0748.png)

**Comment by bridiver** [comment]
why is the page trying to connect to example.com? The only reference I see to that is in a comment

**Comment by srirambv** [comment]
![image](https://cloud.githubusercontent.com/assets/17010094/21654826/288d0f62-d2af-11e6-9c24-c788b68fafd2.png)

**Comment by feross** [comment]
Looks like `example.com` is coming from `stream-http`'s [capability detection code](https://github.com/jhiesey/stream-http/blob/master/lib/capability.js#L13-L17). It doesn't actually make an http request. Just calls `xhr.open()` without `xhr.send()`. But it looks like that still violates the CSP.

@jhiesey - is there any way to remove `example.com` from that code?

**Comment by feross** [comment]
Working on a PR to `stream-http`. Can someone tell me the proper branch to test this on? @bridiver: is `master` still the right place to test?

**Comment by bsclifton** [comment]
@feross master is looking pretty good and is what you should branch from!  :smile: if you haven't recently, please remove node_modules and re-npm install

You can tag me as a reviewer (I'd love to try this functionality out; I haven't tried it yet actually)

**Comment by feross** [comment]
Sounds good! 👍 

Right now, it looks to me like the torrent viewer code isn't even being triggered when clicking on a magnet link. This must have broken at some point and no one noticed. 😐

**Comment by bridiver** [comment]
@feross that is a known issue in 54

**Comment by feross** [comment]
PR to `stream-http` here: https://github.com/jhiesey/stream-http/pull/65

Haven't confirmed this fixes the CSP issue since I haven't gotten the torrent viewer to run on `master` yet.

@bridiver Is there an issue for that?

**Comment by bridiver** [comment]
@feross isn't this the issue?

**Comment by feross** [comment]
@bridiver It looks like `updateWebview` in `frame.js` never gets called when a magnet link is clicked.

That used to get called in the past, and in there we detect it's a `magnet:` url and load the torrent viewer. Any ideas what's preventing that from getting called for `magnet:` links now?

**Comment by bridiver** [comment]
the redirect is handled here https://github.com/brave/browser-laptop/blob/master/app/browser/webtorrent.js#L37 but that never gets hit because the navigation is completely ignored and I'm not sure why yet

**Comment by feross** [comment]
Unlike normal web requests, the magnet link clicks are getting lost here: https://github.com/brave/browser-laptop/blob/b466a44394aeb204bbd5e96f5cef3bc59a8032db/js/components/frame.js#L767 because `e.isMainFrame` is false for some reason. Any ideas?

**Comment by feross** [comment]
Okay, the CSP bug fix was just merged and released in `stream-http`. There should be no more requests to `example.com`. Now, the remaining issue is to fix the breakage that happened when the `chromium54` branch was merged.

@bridiver @bbondy Any ideas why `e.isMainFrame` is `false` for clicks on `magnet:` links but not for normal links. That is causing our handler not to run. See my previous comment.

**Comment by bbondy** [comment]
I'll take a look next, I was trying today but this was blocking me, fixed now:
https://github.com/brave/muon/commit/a1c759260f78c4bf10711f72f54a25d6f1157264

**Comment by bbondy** [comment]
@feross by the way the torrent file downloads ok, but when I try to play it, in this case an MP3 from here:
`magnet:?xt=urn:btih:6a9759bffd5c0af65319979fb7832189f4f3c35d&dn=sintel.mp4&tr=udp%3A%2F%2Fexodus.desync.com%3A6969&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Ftracker.internetwarriors.net%3A1337&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.openbittorrent.com%3A80&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.webtorrent.io&ws=https%3A%2F%2Fwebtorrent.io%2Ftorrents%2Fsintel-1024-surround.mp4`

It errors with this:

<img width="1330" alt="screenshot 2017-01-15 21 05 32" src="https://cloud.githubusercontent.com/assets/831718/21968807/d25c7716-db66-11e6-95bf-c4688a045e45.png">

**Comment by bbondy** [comment]
I got this working even for links inside pages, it'll be in the next release.  I ended up removing the filtering code and doing it through the same API as navigator.registerProtocolHandler uses.
Beyond that I had to adjust `media-src` to accept anything from localhost.

It'll be available in the next preview build (0.13.0-preview10)
